wgpu-core/src/device/queue.rs

```diff
@@ -466,9 +478,96 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
+                > = FastHashMap::default();
+
+                for &cmb_id in command_buffer_ids {
+                    let cmdbuf = command_buffer_guard
+                        .get(cmb_id)
+                        .map_err(|_| QueueSubmitError::InvalidCommandBuffer(cmb_id))?;
+
+                    for buffer_use in cmdbuf.used_buffer_ranges.iter() {
+                        let buffer = buffer_guard
+                            .get_mut(buffer_use.id)
+                            .map_err(|_| QueueSubmitError::DestroyedBuffer(buffer_use.id))?;
+
+                        if let Some(uninitialized_ranges) = buffer
+                            .initialization_status
+                            .drain_uninitialized_ranges(buffer_use.range.clone())
+                        {
+                            match buffer_use.kind {
+                                MemoryInitKind::ImplicitlyInitialized => {
+                                    uninitialized_ranges.for_each(drop);
+                                }
+                                MemoryInitKind::NeedsInitializedMemory => {
+                                    required_buffer_inits
+                                        .entry(buffer_use.id)
+                                        .or_default()
+                                        .extend(uninitialized_ranges);
+                                }
+                            }
+                        }
+                    }
+                }
+                required_buffer_inits
+            };
+
+            if !required_buffer_inits.is_empty() {
+                let (mut device_guard, mut token) = hub.devices.write(&mut token);
+                let (buffer_guard, _) = hub.buffers.read(&mut token);
+                let device = device_guard
+                    .get_mut(queue_id)
+                    .map_err(|_| DeviceError::Invalid)?;
+
+                device
+                    .pending_writes
+                    .dst_buffers
+                    .extend(required_buffer_inits.keys());
+                let pending_writes_cmd_buf = device.borrow_pending_writes();
+
+                for (buffer_id, ranges) in required_buffer_inits {
+                    let buffer_raw = &buffer_guard
+                        .get(buffer_id)
+                        .map_err(|_| QueueSubmitError::DestroyedBuffer(buffer_id))?
+                        .raw
+                        .as_ref()
+                        .ok_or(QueueSubmitError::DestroyedBuffer(buffer_id))?
+                        .0;
+
+                    // TODO: Currently, we implement zero init only via fill_buffer. Should we implement this also via map & shader/storage?
+                    // Note that this requires TRANSFER usage on the buffer - see also device.create_buffer
+
+                    // TODO: Collapse ranges
+                    for range in ranges {
+                        // TODO SHIPSTOPPER: Insert necessary barriers!!
```

Comment:
@kvark Not ready for full review again yet, but here I'd need some pointing into the right direction:

So what's happening here is that we just figured out which ranges need initialization via `fill_buffer`. `device.pending_writes` seems to be the right place to put it. But ofc I also need to create memory barriers and inform the tracker about this.. and this is where I'm struggling a bit. In other places this seems to be done without the tracker? Ideally I'd like to treat this just like another command in a wgpu command-buffer and don't operate on the lowlevel...

(note that this is the _only_ way initialization that I allow right now .. which has the awful implication that the TRANSFER bit needs to be always available :(( see also https://github.com/gfx-rs/wgpu/pull/1159/files#diff-90e54105789cf7268e0acbfd914035fccb5967895c81d21ac60f94af6b439edaR505)

Reply:
> In other places this seems to be done without the tracker? 

Check out the `queue_write_buffer`. It needs exactly the same barriers (which are produced from the tracker):
```rust
let (dst, transition) = trackers
            .buffers
            .use_replace(&*buffer_guard, buffer_id, (), BufferUse::COPY_DST)
            .map_err(TransferError::InvalidBuffer)?;
        unsafe {
            stage.cmdbuf.pipeline_barrier(
                super::all_buffer_stages()..hal::pso::PipelineStage::TRANSFER,
                hal::memory::Dependencies::empty(),
                iter::once(hal::memory::Barrier::Buffer {
                    states: hal::buffer::Access::HOST_WRITE..hal::buffer::Access::TRANSFER_READ,
                    target: &stage.buffer,
                    range: hal::buffer::SubRange::WHOLE,
                    families: None,
                })
                .chain(transition.map(|pending| pending.into_hal(dst))),
            );
            stage
                .cmdbuf
                .copy_buffer(&stage.buffer, dst_raw, iter::once(region));
        }
```

> which has the awful implication that the TRANSFER bit needs to be always available

This is totally fine. For buffers, the set of allowed usages doesn't really make a difference for driver/HW. There is no downside of just internally adding `hal::buffer::Usage::TRANSFER_DST`.

Reply:
perfect that's exactly the hint I needed

> For buffers, the set of allowed usages doesn't really make a difference for driver/HW
would have guess so that usages are only _really_ an issue for textures (because it may change what layout/compression it can do), but I'm very relieved you confirm this :). Then I can remove the TODOs around that aspect